File: sentinel/src/checks/secrets-in-input.ts
/**
 * Secrets-in-input check.
 *
 * Runs on PREFLIGHT (user input → model). Catches API keys, credentials,
 * and connection strings being sent INTO the model prompt — a different
 * threat from `pii-leak` (which catches PII coming OUT).
 *
 * Threat model:
 *   - User pastes their `.env` file contents into a chat to get help
 *     debugging — the AI provider now sees production credentials.
 *   - Developer asks "why is this code failing" and pastes a snippet
 *     containing a hardcoded API key.
 *   - Tool output containing secrets is appended to the user message
 *     (an RAG / agent-tool failure mode).
 *
 * Patterns mirror `lictor-core/src/checks/secrets.rs`. Single source of
 * truth lives in the Rust crate's `SECRET_PATTERNS`; this is a deliberate
 * port for the TypeScript runtime path. Keep both in sync — when one
 * gets a new pattern, the other does too.
 *
 * Privacy: the matched value is fingerprinted (sha256, 16 hex chars) for
 * Guardian telemetry — never shipped raw.
 */

import type { Phase, Severity } from "../types.js";
import type { Check, CheckResult } from "../check-runner.js";
import { PASS } from "../check-runner.js";

export interface SecretPattern {
  re: RegExp;
  label: string;
  severity: Severity;
}

/**
 * Ported from lictor-core/src/checks/secrets.rs::SECRET_PATTERNS.
 * Keep these in lockstep with the Rust source.
 */
export const SECRET_PATTERNS: SecretPattern[] = [
  { re: /AIza[A-Za-z0-9_\-]{35}/g,                                                       label: "Google API key",                              severity: "high" },
  { re: /sk-ant-api\d{2}-[A-Za-z0-9_\-]{40,}/g,                                          label: "Anthropic API key",                           severity: "critical" },
  // Matches both legacy OpenAI (sk-...) and 2024+ project keys (sk-proj-...).
  { re: /sk-(?:proj-)?[A-Za-z0-9_\-]{20,}/g,                                             label: "OpenAI API key (or similar sk- token)",       severity: "critical" },
  { re: /sk_live_[A-Za-z0-9]{24,}/g,                                                     label: "Stripe live secret key",                      severity: "critical" },
  { re: /sk_test_[A-Za-z0-9]{24,}/g,                                                     label: "Stripe test secret key (still leaks logic)",  severity: "medium" },
  { re: /pk_live_[A-Za-z0-9]{24,}/g,                                                     label: "Stripe live publishable key (informational)", severity: "info" },
  { re: /ghp_[A-Za-z0-9]{36}/g,                                                          label: "GitHub personal access token",                severity: "critical" },
  { re: /ghs_[A-Za-z0-9]{36}/g,                                                          label: "GitHub server token",                         severity: "critical" },
  { re: /xox[abp]-[A-Za-z0-9-]{10,}/g,                                                   label: "Slack token",                                 severity: "high" },
  { re: /AKIA[0-9A-Z]{16}/g,                                                             label: "AWS access key ID",                           severity: "high" },
  { re: /-----BEGIN (?:RSA|EC|OPENSSH) PRIVATE KEY-----/g,                                label: "Private key block",                           severity: "critical" },
  { re: /eyJ[A-Za-z0-9_\-]{8,}\.[A-Za-z0-9_\-]{8,}\.[A-Za-z0-9_\-]{8,}/g,                label: "JWT token (verify intended)",                 severity: "low" },
  { re: /mongodb(?:\+srv)?:\/\/[^\s"'<>]+/g,                                             label: "MongoDB connection string",                   severity: "critical" },
  { re: /postgres(?:ql)?:\/\/[^\s"'<>]+/g,                                               label: "PostgreSQL connection string",                severity: "critical" },
  { re: /redis:\/\/[^\s"'<>]+/g,                                                         label: "Redis connection string",                     severity: "high" },
];

const SEVERITY_RANK: Record<Severity, number> = {
  info: 0,
  low: 1,
  medium: 2,
  high: 3,
  critical: 4,
};

function maxSeverity(a: Severity, b: Severity): Severity {
  return SEVERITY_RANK[a] >= SEVERITY_RANK[b] ? a : b;
}

function runSecretsInInput(text: string, _phase: Phase): CheckResult {
  if (!text || text.length === 0) return PASS;

  // Dedup by exact matched value — same secret matching multiple patterns
  // (Anthropic key also matches the broader OpenAI `sk-` pattern) should
  // produce one finding under the more specific label. The order of
  // SECRET_PATTERNS determines which label wins; Anthropic precedes
  // OpenAI in the array so it wins for `sk-ant-api...` strings.
  const seen = new Set<string>();
  const matches: Array<{ label: string; severity: Severity }> = [];
  let topSeverity: Severity = "info";

  for (const pat of SECRET_PATTERNS) {
    pat.re.lastIndex = 0;
    let m: RegExpExecArray | null;
    while ((m = pat.re.exec(text)) !== null) {
      const value = m[0];
      if (seen.has(value)) continue;
      seen.add(value);
      matches.push({ label: pat.label, severity: pat.severity });
      topSeverity = maxSeverity(topSeverity, pat.severity);
      if (matches.length >= 10) break;
      if (!pat.re.global) break;
    }
    if (matches.length >= 10) break;
  }

  if (matches.length === 0) return PASS;

  const labels = matches.map((m) => `  [${m.severity}] ${m.label}`).join("\n");
  const title =
    matches.length === 1
      ? `Secret in input: ${matches[0]!.label}`
      : `Secrets in input — ${matches.length} distinct credentials`;
  const detail =
    `Pattern-based detection of hardcoded credentials in the user's input.\n` +
    `Sending these to the model leaks them to the AI provider (logged, ` +
    `potentially trained on, or visible to provider support staff).\n\n` +
    `Matched:\n${labels}\n\n` +
    `Remediation: the secret should be redacted before the prompt is sent. ` +
    `If using @lictor/sentinel with onTrip: 'redact', secrets are stripped ` +
    `automatically. With onTrip: 'block', the request is rejected entirely. ` +
    `With onTrip: 'log' (default), the event is recorded but the call proceeds — ` +
    `you should rotate the leaked credential.`;

  return {
    tripped: true,
    severity: topSeverity,
    title,
    detail,
  };
}

export const secretsInInputCheck: Check = {
  id: "secrets-in-input",
  run: runSecretsInInput,
};

File: sentinel-py/src/lictor_sentinel/checks/secrets_in_input.py
"""
Secrets-in-input check (Python) — preflight. Mirrors
@lictor/sentinel/src/checks/secrets-in-input.ts and lictor-core's
SECRET_PATTERNS in Rust.

When a pattern is added to one of those, mirror it here.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

from ..check_runner import PASS, Check, CheckResult
from ..types import Phase, Severity


@dataclass(frozen=True)
class SecretPattern:
    re: re.Pattern[str]
    label: str
    severity: Severity


SECRET_PATTERNS: list[SecretPattern] = [
    SecretPattern(re.compile(r"AIza[A-Za-z0-9_\-]{35}"),                                                     "Google API key",                              "high"),
    SecretPattern(re.compile(r"sk-ant-api\d{2}-[A-Za-z0-9_\-]{40,}"),                                         "Anthropic API key",                           "critical"),
    SecretPattern(re.compile(r"sk-(?:proj-)?[A-Za-z0-9_\-]{20,}"),                                            "OpenAI API key (or similar sk- token)",       "critical"),
    SecretPattern(re.compile(r"sk_live_[A-Za-z0-9]{24,}"),                                                    "Stripe live secret key",                      "critical"),
    SecretPattern(re.compile(r"sk_test_[A-Za-z0-9]{24,}"),                                                    "Stripe test secret key (still leaks logic)",  "medium"),
    SecretPattern(re.compile(r"pk_live_[A-Za-z0-9]{24,}"),                                                    "Stripe live publishable key (informational)", "info"),
    SecretPattern(re.compile(r"ghp_[A-Za-z0-9]{36}"),                                                         "GitHub personal access token",                "critical"),
    SecretPattern(re.compile(r"ghs_[A-Za-z0-9]{36}"),                                                         "GitHub server token",                         "critical"),
    SecretPattern(re.compile(r"xox[abp]-[A-Za-z0-9\-]{10,}"),                                                 "Slack token",                                 "high"),
    SecretPattern(re.compile(r"AKIA[0-9A-Z]{16}"),                                                            "AWS access key ID",                           "high"),
    SecretPattern(re.compile(r"-----BEGIN (?:RSA|EC|OPENSSH) PRIVATE KEY-----"),                              "Private key block",                           "critical"),
    SecretPattern(re.compile(r"eyJ[A-Za-z0-9_\-]{8,}\.[A-Za-z0-9_\-]{8,}\.[A-Za-z0-9_\-]{8,}"),               "JWT token (verify intended)",                 "low"),
    SecretPattern(re.compile(r"mongodb(?:\+srv)?://[^\s\"'<>]+"),                                             "MongoDB connection string",                   "critical"),
    SecretPattern(re.compile(r"postgres(?:ql)?://[^\s\"'<>]+"),                                               "PostgreSQL connection string",                "critical"),
    SecretPattern(re.compile(r"redis://[^\s\"'<>]+"),                                                         "Redis connection string",                     "high"),
]

_SEVERITY_RANK: dict[Severity, int] = {"info": 0, "low": 1, "medium": 2, "high": 3, "critical": 4}


def _max_severity(a: Severity, b: Severity) -> Severity:
    return a if _SEVERITY_RANK[a] >= _SEVERITY_RANK[b] else b


def _run(text: str, _phase: Phase) -> CheckResult:
    if not text:
        return PASS
    seen: set[str] = set()
    matches: list[tuple[str, Severity]] = []
    top: Severity = "info"

    for pat in SECRET_PATTERNS:
        for m in pat.re.finditer(text):
            value = m.group(0)
            if value in seen:
                continue
            seen.add(value)
            matches.append((pat.label, pat.severity))
            top = _max_severity(top, pat.severity)
            if len(matches) >= 10:
                break
        if len(matches) >= 10:
            break

    if not matches:
        return PASS

    labels = "\n".join(f"  [{sev}] {lbl}" for lbl, sev in matches)
    title = (
        f"Secret in input: {matches[0][0]}"
        if len(matches) == 1
        else f"Secrets in input — {len(matches)} distinct credentials"
    )
    detail = (
        "Pattern-based detection of hardcoded credentials in the user's input.\n"
        "Sending these to the model leaks them to the AI provider (logged, "
        "potentially trained on, or visible to provider support staff).\n\n"
        f"Matched:\n{labels}\n\n"
        "Remediation: the secret should be redacted before the prompt is sent. "
        "With on_trip='block', the request is rejected entirely. "
        "With on_trip='log' (default), the event is recorded but the call proceeds — "
        "rotate the leaked credential."
    )
    return CheckResult(tripped=True, severity=top, title=title, detail=detail)


class _SecretsInInputCheck:
    id = "secrets-in-input"

    def run(self, text: str, phase: Phase) -> CheckResult:
        return _run(text, phase)


secrets_in_input_check: Check = _SecretsInInputCheck()  # type: ignore[assignment]
